src/core/Akka/Dispatch/MessageQueues/UnboundedPriorityMailboxQueue.cs

```diff
@@ -12,63 +12,63 @@
 namespace Akka.Dispatch.MessageQueues
 {
     /// <summary> 
-    /// Base class message queue that uses a priority generator for messages 
+    /// Base class for a message queue that uses a priority generator for messages 
     /// </summary>
     public class UnboundedPriorityMessageQueue : BlockingMessageQueue
     {
         private readonly ListPriorityQueue _prioQueue;
 
         /// <summary>
-        /// TBD
+        /// DEPRECATED. Use UnboundedPriorityMessageQueue(Func{object, int} priorityGenerator, int initialCapacity) instead.        
         /// </summary>
-        /// <param name="initialCapacity">TBD</param>
-        public UnboundedPriorityMessageQueue(int initialCapacity)
+        /// <param name="initialCapacity">The initial capacity of the priority queue.</param>
+        [Obsolete("Use UnboundedPriorityMessageQueue(Func<object, int> priorityGenerator, int initialCapacity) instead.")]
+        public UnboundedPriorityMessageQueue(int initialCapacity) : this(ListPriorityQueue.DefaultPriorityCalculator, initialCapacity)
         {
-            _prioQueue = new ListPriorityQueue(initialCapacity);
-        }
 
-        /// <summary>
-        /// TBD
-        /// </summary>
-        /// <param name="priorityGenerator">TBD</param>
-        /// <param name="initialCapacity">TBD</param>
-        public UnboundedPriorityMessageQueue(Func<object, int> priorityGenerator, int initialCapacity) : this(initialCapacity)
-        {
-            _prioQueue.SetPriorityCalculator(priorityGenerator);
         }
 
         /// <summary>
-        /// TBD
+        /// Creates a new unbounded priority message queue.
         /// </summary>
-        /// <param name="priorityGenerator">TBD</param>
-        /// <returns>TBD</returns>
-        internal void SetPriorityGenerator(Func<object, int> priorityGenerator)
+        /// <param name="priorityGenerator">The calculator function for determining the priority of inbound messages.</param>
+        /// <param name="initialCapacity">The initial capacity of the queue.</param>
+        public UnboundedPriorityMessageQueue(Func<object, int> priorityGenerator, int initialCapacity)
         {
-            _prioQueue.SetPriorityCalculator(priorityGenerator);
+            _prioQueue = new ListPriorityQueue(initialCapacity, priorityGenerator);
```

Comment:
This is the key fix: create the queue WITH its priority generator at the same time and never change it again. That way we don't end up in a situation where the mailbox is having messages queued into it while the underlying priority generator is being changed concurrently on another thread.